client/client.go

```diff
@@ -18,173 +18,147 @@ import (
-		return err
-	}
-	if dec.More() {
-		return fmt.Errorf("cannot parse json value")
-	}
-	return nil
-}
-
-func unixDialer(socketPath string) func(string, string) (net.Conn, error) {
-	return func(_, _ string) (net.Conn, error) {
-		_, err := os.Stat(socketPath)
-		if errors.Is(err, os.ErrNotExist) {
-			return nil, &SocketNotFoundError{Err: err, Path: socketPath}
-		}
-		if err != nil {
-			return nil, fmt.Errorf("cannot stat %q: %w", socketPath, err)
-		}
-
-		return net.Dial("unix", socketPath)
-	}
-}
-
-type doer interface {
-	Do(*http.Request) (*http.Response, error)
-}
-
-// Config allows the user to customize client behavior.
-type Config struct {
-	// BaseURL contains the base URL where the Pebble daemon is expected to be.
-	// It can be empty for a default behavior of talking over a unix socket.
-	BaseURL string
-
-	// Socket is the path to the unix socket to use.
-	Socket string
-
-	// DisableKeepAlive indicates that the connections should not be kept
-	// alive for later reuse (the default is to keep them alive).
-	DisableKeepAlive bool
-
-	// UserAgent is the User-Agent header sent to the Pebble daemon.
-	UserAgent string
+// DecoderFunc allows the client access to the HTTP response. See the SetDecoder
+// description for more details.
+type DecoderFunc func(ctx context.Context, res *http.Response, opts *RequestOptions, result interface{}) (*RequestResponse, error)
+
+type Requester interface {
+	// Do must support the following cases:
+	//
+	// 1. Sync response
+	// 2. Async response
+	// 3. Error response
+	// 4. Websocket creation request
+	// 5. Raw HTTP body request
+	//
+	// See the default implementation for information on what is expected in
+	// each use case.
+	Do(ctx context.Context, opts *RequestOptions, result interface{}) (*RequestResponse, error)
+
+	// SetDecoder allows for client specific processing to be hooked into
+	// the sync and async response decoding process. The decoder is also
+	// responsible for unmarshalling the result, and populating the
+	// RequestReponse.
+	SetDecoder(decoder DecoderFunc)
+}
+
+// RequestOptions allows setting up a specific request.
+type RequestOptions struct {
+	Method  string
+	Path    string
+	Query   url.Values
+	Headers map[string]string
+	Body    io.Reader
+	Async   bool
+}
+
+// RequestResponse defines a common response associated with requests.
+type RequestResponse struct {
+	StatusCode int
+	Type       string
+	Change     string
 }
 
 // A Client knows how to talk to the Pebble daemon.
 type Client struct {
-	baseURL   url.URL
-	doer      doer
-	userAgent string
-
-	maintenance error
+	Requester Requester
 
+	maintenance      error
 	warningCount     int
 	warningTimestamp time.Time
-
-	getWebsocket getWebsocketFunc
 }
 
-type getWebsocketFunc func(url string) (clientWebsocket, error)
-
-type clientWebsocket interface {
-	wsutil.MessageReader
-	wsutil.MessageWriter
-	io.Closer
-	jsonWriter
+func New(requester Requester) (*Client, error) {
```

Comment:
I like this better than having a separate `NewWithRequester()` function, however—is it acceptable that we break the existing client API?

Reply:
Fixed.